Create Windows-Terminal-Integration.md

capocasa opened the pull request:
I was having trouble installing build dependencies to build a portable, unixy library (dav1d) for the Windows terminal (my build system needs that), so I came up with this. I think it's very exciting, it really expands the possible use cases for msys2 and seemed to work really well. I also didn't see it used this way so far, everyone seems to think msys2 is just for unix users.

mingwandroid commented:
I use MSYS2 in Windows Terminal, but in a very different way to you. I want to use the MSYS2 shell(s), not cmd.exe (does anyone want to use cmd.exe?)

I would be more supportive of this (I don't really mind it FWIW) if it addressed using MSYS2 as a package manager within PowerShell as that is still supported by Microsoft.

To integrate MSYS2 with the bash shell into Microsoft Terminal, I added the following to settings:

```
            {
                "name": "MSYS2",
                "commandline": "cmd.exe /c set MSYS=winsymlinks:nativestrict&& set MSYS2_PATH_TYPE=inherit&& set CHERE_INVOKING=1&& C:/msys64/usr/bin/bash.exe -i -l",
                "guid": "{71c54bbd-c2c6-5271-96e7-009a87ff44b0}",
                "backgroundImage": "ms-appdata:///local/msys2-large-3.png",
                "icon": "C:\\msys64\\msys2.ico"
            },
```

eine commented:
I think the bat file should better look like https://github.com/msys2/setup-msys2/blob/master/main.js#L165-L169, which is equivalent to @mingwandroid's configuration, and it's also the one provided in GitHub Actions' virtual-environments: https://github.com/actions/virtual-environments/pull/2395#issuecomment-757546669. There is a much more detailed `msys2_shell.cmd` in MSYS2 installation.

Most of tools installed through pacman can already be used from powershell by just executing them. I think that's the most important feature to explain. See, for instance: https://github.com/ghdl/setup-ghdl-ci/blob/master/.github/workflows/test.yml#L68-L89. IMHO, setting a custom envvar, such as MSYS2_PATH is desirable instead of adding MSYS2 to the PATH.

Other than that, I find this PR very interesting and useful.

lazka commented:
> To integrate MSYS2 with the bash shell into Microsoft Terminal, I added the following to settings:

We also have a page for this btw: https://www.msys2.org/docs/terminals/

@capocasa I'm unsure which problem you are trying to solve. We build dav1d with meson without any unix tools (meson has its own shebang interpreter if the need arises). And if you want to build with autotools why not use bash as well?

capocasa commented:
@mingwandroid, @eine, @lazka

Thank you for your feedback and your questions! I'm sorry, I was sloppy with the definitions of Terminal and Shell, which I believe has resulted in some confusion. I updated my pull request to be more clear, does it make more sense now?

I am aware that the msys2-bash shell can be run within the Windows Terminal, which is one level of integration.

Another level of integration is to run msys2-installed programs within cmd or PowerShell, without bash. That's already being done all the time for various reasons, but the users are Frankensteining together their own suites of unix commands- bash and git from the git package, ssh from openssh, and so on. This is pretty terrible- you never know when you'll hit issues, if you can find the commands you need at all. I was doing the same thing, until I started unpacking msys2 packages to get commands I couldn't find elsewhere. They worked so well that it occurred to me to just install msys2-pacman and put the installed packages in the path. The reason I didn't try msys2 earlier is because I had read on the "scoop" web site that "msys2 is a unix subsystem, not just a package manager", which sounded heavy and inflexible. But I found I didn't need to use everything, I could just use the package manager and it just works.

The reason I needed to do that is because the automation tool I was using `cmd` to run its external commands. Sure, I could send the author a pull request, but I'm going to run into another one with similar limitations soon enough. Just putting the msys2 programs in the path is a heck of a lot less friction than arguing with various people about which shells they should support in their software.

I think a lot of people who don't want to use a "unix on windows" or use bash absolutely want a consistent set of unix tools, and I created the page in the hopes of sharing this information with them because it made my life so much easier.

@mingwandroid My instructions work equally well in cmd and PowerShell, I changed them to better reflect that. I can't use msys2-bash because the automation tool runs my commands for me in `cmd`.

@eine Can you tell me why CHERE_INVOKING=1, setlocal and MSYS2_PATH_TYPE are necessary? I'm trying to keep things as simple as possible. msys2_shell.cmd doesn't match my terminal, the wrapper should only execute a script within cmd or PowerShell, not open a bash terminal Window.

@lazka The problem I want to solve is running msys2 packages from `cmd` to humor a build system that requires it, and to replace worse partial unix setups that already do this.

eine commented:
> @eine Can you tell me why CHERE_INVOKING=1, setlocal and MSYS2_PATH_TYPE are necessary? I'm trying to keep things as simple as possible. msys2_shell.cmd doesn't match my terminal, the wrapper should only execute a script within cmd or PowerShell, not open a bash terminal Window.

You need to clarify your scope. There are two different issues here. On the one hand, you want to run Unix tools installed through MSYS2's pacman in the cmd or powershell. On the other hand, you want to wrap MSYS2's shell in a custom entrypoint to bash. 

For the first point, it does already work without tweaks. On Windows, the PATH is unnatural. Such concept does not exist. All users are expected to execute each tool by providing the location explicitly. Therefore, if your target are users which don't want to be exposed to Unix things, adding anything to the PATH should not be considered. You can provide MSYS2_PATH, for them not to know/remember the location of the MSYS2 installation. That would be the natural procedure for a Windows only experience. Those are the Git for Windows and Powershell examples I referred.

For the second point, MSYS2's bash should not be executed by calling `bash` only. There are three different MSYS2 shells: MSYS, MINGW32 and MINGW64. Moreover, by default, shells will start in a fixed location, and the PATH will be strictly constrained so that MSYS2's tools are visible only.

- MSYSTEM allows selecting the type of terminal/environment.
- CHERE_INVOKING allows modifying the location/workdir.
- MSYS2_PATH_TYPE allows modifying which external tools/envvars are visible in the shell.

See:

- https://github.com/msys2/setup-msys2#msystem
- https://github.com/msys2/msys2.github.io/blob/source/docs/docs/ci.md
- https://github.com/msys2/setup-msys2#path-type

My concern is that providing a usable `myshellscript.bat` of `myshellscript.cmd` makes your first scope negligible. If you expect users to be aware of some Unix things, such as a PATH and a list of binaries that can be found through it, you might well educate them for prepending `myshellscript -c` to the Unix tools they want to use. That will work not only from cmd and powershell, but also from virtually any other tools on Windows. For example, say executing a shell script from Matlab. `system('myshellscript', '-c', ...)` works.

Overall, the question to be documented here is: how do users execute a sequence of commands on an MSYS2 shell, without graphically opening an MSYS2 window? Technically, the answer is the same as for CI (see https://github.com/msys2/msys2.github.io/blob/source/docs/docs/ci.md or setup-msys2). Therefore, I think that this PR is currently unnecessarily complex/complicated. Yet, that is a natural procedure when you are still wrapping your head around a new topic. I think it could be rethought so that the main motto to the user is: "*don't panic! you don't need to do anything, just pay attention and understand*". The following three quotes are a good starting point: https://hdl.github.io/MINGW-packages/#_usage.

> That's already being done all the time for various reasons, but the users are Frankensteining together their own suites of unix commands- bash and git from the git package, ssh from openssh, and so on. This is pretty terrible- you never know when you'll hit issues, if you can find the commands you need at all. I was doing the same thing, until I started unpacking msys2 packages to get commands I couldn't find elsewhere. 

This is mostly lack of experience. All the users that did so in CI and were suggested to use setup-msys2 instead, did significantly reduce their maintenance burden and did achieve a much better understanding of the environment. If you do need tools from MSYS2 for your build process, that means that the environment outside of MSYS2 is broken already. Trying it to keep external is fighting the evidence.

mingwandroid commented:
> I think a lot of people who don't want to use a "unix on windows" or use bash absolutely want a consistent set of unix tools

I don't really agree on this point, but "a lot of people" is pretty vague, I'd argue that this is a small group, relative to our user size (massive when you also add in Git for Windows and conda on Windows users). MSYS2 proves that "Unix-ish on Windows" can be lightweight. A basic install is pretty small. I also think most people who want a consistent set of unix tools will also want to use bash (esp if the alternative is to script in bash). The MSYS2/mingw-w64 tools work fine outside the Unix shell anyway.

> Another level of integration is to run msys2-installed programs within cmd or PowerShell, without bash. That's already being done all the time for various reasons, but the users are Frankensteining together their own suites of unix commands- bash and git from the git package, ssh from openssh, and so on

These packaging issues for 3rd parties are easily solved, in fact getting Windows developers away from Frankensteining random bits together in the hope of getting some software that works is one of my main motivations for being involved in MSYS2.

Packaging Frankenstien software is pointless, packaging MSYS2 software is easy, especially if you package you own stuff with pkgbuild and declare its dependencies correctly. Then to package all you do is to create a new root with pacman and install your package into it, then you make an installer with whatever installer tech you like. MSYS2 itself uses The Qt Installer Framework. Our scripts are available if anyone wishes to use them. Git for Windows uses something else. Here's one approach: https://blogs.gnome.org/nacho/2014/08/01/how-to-build-your-gtk-application-on-windows/

capocasa commented:
@eine

Thank you for your explanations!

> On the one hand, you want to run Unix tools installed through MSYS2's pacman in the cmd or powershell

This is correct. 

> you want to wrap MSYS2's shell in a custom entrypoint to bash

This is not correct- I don't want to enter the msys2 shell at all. I want to execute scripts that come with msys2, such as aclocal, with whatever inter

> makes your first scope negligible. If you expect users to be aware of some Unix things, such as a PATH

It does not, there is no contradiction here- PATH is not a unix exclusive construct, it has been around in Windows shells and all the way back to DOS times. Giving the user somewhat detailed instructions about his or her existing system is fine, merely foreign-looking instructions like bash syntax or unix-style paths are not.

> You need to clarify your scope

Yes, gladly: I have a build tool that will execute the commands `bash customScript`, `autoreconf` and `make`,  in the `cmd` shell, to which `cmd` replies with an error message. I want `cmd` to execute those commands.

Placing the msys2 programs into the windows path works for the commands that are executables, and adding an corresponding .bat file that will run the interpreter works for shell scripts. It could be prettier but it does work. I think it would make sense to share these instructions for those who need them as a useful "unofficial" way of using msys2. I can gladly change the instructions if better ones that also work come up.

eine commented:
> > you want to wrap MSYS2's shell in a custom entrypoint to bash
> 
> This is not correct- I don't want to enter the msys2 shell at all.

You do want to do so. You keep using different words to explain it, but it's the same technical solution:

> and adding an corresponding .bat file that will run the interpreter works for shell scripts

"*Adding a corresponding .bat file that will run the interpreter*" is exactly the same as "*Providing a custom entrypoint to bash*". The difference is the extension, you can use `*.bat` or `*.cmd`.

> Giving the user somewhat detailed instructions about his or her existing system is fine,

The detailed instructions about this should be: do never add the location of MSYS2 binaries to the PATH blindly; DLL issues might and will arise. Expose only the Unix tools that you need to use outside. Preferredly, call everything from an initial bash shell, so that all the tools are exposed for you without conflicts. Calling cmd or powershell from bash works better than the opposite. 

> merely foreign-looking instructions like bash syntax or unix-style paths are not.

The whole purpose is using Unix tools and running the interpreter (bash). The syntax of the paths is the less strange feature users will need to deal with. All the CLI options will use different syntax and constraints, because most Unix tools are close to being POSIX compatible and do follow GNU guidelines.

capocasa commented:
Thank you for patiently explaining!

> it's the same technical solution

I see- I thought you were referring to instructions to open an msys2 terminal window.

> never add the location of MSYS2 binaries to the PATH blindly, DLL issues might and will arise

Fair enough.

> whole purpose is using Unix tools and running the interpreter

I understand that this is the stated goal of the project, I just noticed the project can be used in an unintended way to solve the problem of "I just need a few unix executables in my Windows shell", and it works better than any other solution I found so far.

> call everything from an initial bash shell

I understand that using-as-intended works better, but as mentioned this won't work in this case- my automated tool can't depend on msys2 explicitly, it needs to run the programs like they were regular windows programs.

> Expose only the Unix tools that you need to use outside

Fair enough. Phrased without supposing msys2 knowledge: Do I understand you correctly that, if I don't want to use msys2 programs safely form outside msys2, my only option is to create a .bat/.cmd file in the windows path that will then launch the corresponding msys2 executable using msys2 bash with appropriate arguments, e.g `C:\\msys64\\usr\\bin\\env MSYSTEM=MSYS /usr/bin/bash -li.`?

I'm acting so surprised about this even though it's documented because when I just tried directly calling the executables by full path, even copying them outside C:\msys64\usr\bin and running them from elsewhere, it worked without a hitch for the sub-set of exectables I tried it with.

eine commented:
> I understand that this is the stated goal of the project, I just noticed the project can be used in an unintended way to solve the problem of "I just need a few unix executables in my Windows shell", 

Those "unix executables" are "native Windows executables" indeed. Those tools are coming from the Unix ecosystem, but are not Unix executables once they were built on MINGW32/MINGW64 shells. Therefore, you can use them outside of MSYS2's shells.

> I'm acting so surprised about this even though it's documented because when I just tried directly calling the executables by full path, even copying them outside C:\msys64\usr\bin and running them from elsewhere, it worked without a hitch for the sub-set of exectables I tried it with. So I guess my follow-up question would be: if this is shouldn't be done, why does it work so well?

In fact, you can pick MSYS2 packages (ZST tarballs), extract them, generate a zipfile and distribute that to your users (or through Chocolatey). That will work without any issue, as long as the executables are statically compiled. Otherwise, the content of the ZST tarballs alone won't work, you will need to add the dependencies into the zipfile. That can be done manually. For instance, in [gtkwave/gtkwave](https://github.com/gtkwave/gtkwave), that's how standalone windows zipfiles are generated. In https://github.com/ghdl/ghdl/issues/1560#issuecomment-752372152 a user reported using PyCharm/CPython along with binaries and DLLs extracted from an MSYS2 installation. However, as explained in https://github.com/ghdl/ghdl/issues/1560#issuecomment-752750494, the process for generating a standalone zipfile from an MSYS2 package is not automated yet.

IMHO, since MSYS2 provides pacman, and all of those tools need to be built on MSYS2, it is pointless to try hiding it from the user. Anything other than using pacman is reinventing the wheel, unless you have very specific version pinning requirements which conflict with the rolling nature. Assuming that the less painful install procedure for most users is getting MSYS2 and using pacman, I'd expect most users to have a working MSYS2 installation and not a mutilated variant.

> and it works better than any other solution I found so far.

> I understand that using-as-intended works better, but as mentioned this won't work in this case- my automated tool can't depend on msys2 explicitly

Using MSYS2 in the intended way works better than any other solution. Effort is better spent on educating your users than trying to develop an unnecessary complexity layer. This is because your automated tool does depend on MSYS2 explicitly, since the tools need to be installed through pacman (unless you Frankenstein DLLs).

> it needs to run the programs like they were regular windows programs.

> if I want to use msys2 programs safely form outside an msys2 bash shell, my only option is to create a .bat/.cmd file in the windows path that will then launch the corresponding msys2 executable using msys2 bash with appropriate arguments, e.g C:\\msys64\\usr\\bin\\env MSYSTEM=MSYS /usr/bin/bash -li.?

https://github.com/ghdl/setup-ghdl-ci/blob/master/.github/workflows/test.yml#L88-L89

Executing binaries directly is no issue because those are regular windows programs. The requirement of a `*.bat`/`*.cmd` applies only to executing bash/shell scripts without graphically opening an MSYS2 window. The issues in that case are:

- The environment variables that define the behaviour of bash.
- The syntax of the locations you pass as arguments. Those need to be Unix paths, because you are explicitly asking it to be interpreted as bash.

You can hide the first one, but the second one is trickier. If users need to be aware of that (or the build tool needs to use `cygpath`), there is no point in the argument that "users don't want/need to be exposed to Unix things". They do need to as long as the tools they are using depend on packages built on MSYS2. Doing so is easier and faster than any other alternative for providing Unix tooling on Windows. Overall, if you wanted to provide a single tool/application, that'd be much easier. But you seem to want multiple tools to be used, which kind of involves rebundling a subset of an MSYS2 installation.

As said, I'm all in with clarifying these issues in the docs. This dialogue proves that it is not easy to understand, even for users who really want to do it properly.

capocasa commented:
@eine thank you yet again!

> Therefore, you can use them outside of MSYS2's shells.

I've got to repeat how great that is. 

> Using MSYS2 in the intended way works better than any other solution

That depends on your use case. For mine, it doesn't work at all.

I understand that when I am suggesting to abuse a finely honed system like msys2 in the way I am suggesting that it might ruffle some feathers- it's certainly not an optimal solution, let alone an aesthetically pleasing one. But consider the alternative, the Frankenstein toolchain- it's much, much worse. In my situation, I simply don't have the luxury of using msys2-bash- there are too many tools, legacy and modern, that, at the moment, call basic unix commands using `cmd.exe` and it's just not feasible to attempt to change that. These systems cannot be educated, and it would be obscenely expensive for the authors to change them. I have to provide the unix commands these systems expect somehow. I can put them in the Windows path and add entrypoints for scripts, I can have a collection of entrypoints in the Windows path for all the tools, or I can have either to the working directory, whatever that is for each tool. All of these options are messy. But it appears to me that the option as described in my writeup ist the least messy. This is not a beauty contest, this is a least-terrible-option contest, and for a the drop-in replacement of Frankenbash that I desperately need, like it or not, you're it.

Can you consider allowing this on your docs as an off-label use case, with a disclaimer that "Your Mileage May Vary"? I think your docs are fine for msys2's intended way of use, I had absolutely no problems finding my way around the normal way of using it. Quite the contrary, I love it. It's just that, whatever issue you may take with what I described, it works to build the stuff I need in the ecosystem I need, and nothing else does.

eine commented:
@capocasa, note that I ain't a maintainer of MSYS2. I'm a co-maintainer of Action setup-msys2. I'm no authority and my opinion is just as valuable as yours. My purpose is sharing my background with you, not deciding what is allowed.

> But consider the alternative, the Frankenstein toolchain- it's much, much worse.

I do agree.

> I simply don't have the luxury of using msys2-bash- there are too many tools, legacy and modern, that, at the moment, call basic unix commands using cmd.exe and it's just not feasible to attempt to change that.

I am very curious about which Windows tools call Unix commands using `cmd.exe`, but cannot be aware of MSYS2's bash. That sounds as a poorly engineered solution. This is not a criticism: many projects just want to prove that the codebase is compilable on Windows but don't have any real interest in providing a decently packaged solution. That's legit. However, translating the problem to MSYS2's side (to the users of those tools, to you) is not the most sensible approach.

> These systems cannot be educated, and it would be obscenely expensive for the authors to change them.

Then, maybe they should not claim Windows to be supported. Or they should use a traditional VS based approach and deal with replacing Unix tools with other Windows equivalents. I'm being very cheeky and rude here, as a dialectical trick. But I hope you get what I mean.

> I have to provide the unix commands these systems expect somehow. I can put them in the Windows path and add entrypoints for scripts, I can have a collection of entrypoints in the Windows path for all the tools, or I can have either to the working directory, whatever that is for each tool. All of these options are messy.

You forgot the most straightforward solution: call whichever Windows tool you want to use, from MSYS2's bash. If you open a bash shell first and then call any tool from there (any native Windows tool which might have been built and distributed completely outside of MSYS2), that one inherits the environment. Problem solved. Your tool will find all the Unix tools per se, without any tweak, any custom entrypoints...

> But it appears to me that the option as described in my writeup ist the least messy. This is not a beauty contest, this is a least-terrible-option contest, and for a the drop-in replacement of Frankenbash that I desperately need, like it or not, you're it.

Adding `C:\msys64\usr\bin` and `C:\msys64\mingw64\bin` to the PATH is the most terrible option. If that was sensible, MSYS2's path visibility options would not exist. It exists, and it defaults to `strict` (not `inherit`) for a reason. Don't take me wrong: I wish MSYS2 could share the full PATH with Windows; that'd be awesome for all of us; but reality is different.

There are alternatives to the Frankenbash, but any of them requires understanding the tools that one is using: e.g. entering bash first and calling the tool from there. This can be done in the shortcut execution command. I believe you can make it easier for users and maintainers of third-party projects to understand, but any solution such as "copy this two values to the path and you are ready to go" is fundamentally broken. Note that's very different from "all binaries built/installed in MSYS2 can be executed as any other native tool". They might sound similar, but the very relevant difference is the environment.

> Can you consider allowing this on your docs as an off-label use case, with a disclaimer that "Your Mileage May Vary"?

IMHO, adding `C:\msys64\usr\bin` and `C:\msys64\mingw64\bin` to the PATH is unacceptable in MSYS2's docs, unless it is for explicitly explaining what does not work. Even if done, the order is wrong: in case of duplication, you need mingw32/mingw64 to be found before the MSYS versions. Moreover, you will have conflicts with `tar` and other tools available in `C:\Windows\System32`, unless the MSYS2 paths are placed at the beginning of the PATH, not at the end. If you do so, all the Windows tools which do indeed depend on the Windows versions will fail, because MSYS2's will be found first. If a user is proficient enough for understanding this, wrapping their head about what MSYS2 is won't be difficult.

A year ago, some maintainer of GitHub's virtual environments decided to add Git for Windows' internal Unix tools to the PATH by default. Git for Windows' internal Unix tools are a mutilated subset of MSYS2. You can see in actions/virtual-environments#1525 all the issues and conflicts derived from that and when trying to do the same with MSYS2.

Anyway, I repeat I am not a maintainer and I have no take on this. I'm just sharing my experience and strongly suggesting to evaluate the maintainance burden that you are putting on yourself by documenting that approach and providing it to your users.

> It's just that, whatever issue you may take with what I described, it works to build the stuff I need in the ecosystem I need, and nothing else does.

Please, try entering bash first, and then executing your build/ecosystem entrypoint from there. You can use the `*.cmd` file from setup-msys2 for: msys2.cmd -> bash -> cmd.exe build entrypoint.

> So here's what happens- someone writes a package on Linux or BSD, and decides to use a a few shell commands because it's easy or the language doesn't provide anything similar. Then someone ports it to OSX, everyone's happy. By that time, this package has a user base, it's hard to change things. Then someone wants to port it to Windows, and stumbles over the missing commands- and there you have the incentive to do the Frankenshell. At this point, as a library porter, you're certainly not going to replicate the shell commands using language facilities and risk blowing things up for all the library's existing users, and you're kind of annoyed and think it's not worth your time to do this in the first place. So you either Frankenshell, or the package goes unported on Windows. This happens over and over and over again- I've seen it happen so many times it's just sad.

I do understand the use case perfectly, as I've been there dozens of times. Yet, as sad as it can be, it's how economics and computer sciences work. If you want to port a tool from Unix to Windows, you need expertise (both about the tool and about Windows), it will likely be expensive and sometimes not possible. Windows decided not to support open and standard protocols for 30-40 years because reasons. Then, they decided to go for WSL, apparently ignoring MSYS2. You are using Windows for development because reasons. Pay for it, use WSL or deal with it... that's what the company charging for your OS is offering and/or what your employer/clients are forcing you to do. MSYS2 is an alternative to that reality, which provides a much easy going workflow that produces usable binaries; but it comes with some very few and specific constraints. It is up to you:

1. Fund the projects or contribute for supporting Windows properly.
2. Fund the projects or contribute for supporting MSYS2 as a build environment properly.
3. Use WSL.
4. Use Linux containers (either Docker Desktop or WSL2).
5. Use Linux virtual machines.
6. Use MSYS2.

My preference is 6-2-4. First I help use/package the tools on MSYS2, if available. Else, I try contributing upstream so that building on MSYS2 is supported properly. Else, I use a Linux container, which does always work for Unix tools.

> I'm going out on this tangent to demonstrate how great the need is for direct cmd.exe support for running msys2 programs, which hopefully makes it more understandable why I am willing to look into the rather unclean solution I am proposing. I am indeed rather desperate for a solution.

Honeslty, `cmd.exe` is or should be ditched. It would be much easier if you explained which specific tools you need/want to use and which are the dependencies you are having issues with. If the only problem you have is how to find some Unix tools, chances are that your tool can be handled through a PKGBUILD recipe (which Ray suggested above). That will make everything offensively easy for everyone (you, us, your users...). The discussion above, your intent and ending with adding some PKGBUILD to MSYS2 is the most common and natural procedure for MSYS2 contributors. Yet, docs feel like written for users who want to be contributors in first place. It does not properly explain that contributing a PKGBUILD recipe is the easiest and most sensible approach as soon as you start thinking about writting custom entrpoints.

capocasa commented:
> Please, try entering bash first, and then executing your build/ecosystem entrypoint from there

You're just not listening to me. I cannot do this.

> I am very curious about which Windows tools call Unix commands using cmd.exe

Try python

eine commented:
> You're just not listening to me. I cannot do this.

You can write a `*.bat` or `*.cmd` file that does it for you. How do you execute the tool normally?

> Try python

That's a interesting one!

- https://github.com/msys2/MINGW-packages/discussions/7619
- https://github.com/msys2/MINGW-packages/discussions/7646

Better pick any other example.

capocasa commented:
I'm sorry, but I wrote extensively how I execute the tool- from the "system" command from programming languages, and those use the default shell. This is also the case for python, I just tried it.

eine commented:
Precisely, I'm suggesting to write a `*.bat`/`*.cmd` file, which is to be found by the system command, instead of the "real" tool. Still, I'd need to know which system command is that. For a Python shell, the en:

```bat
@echo off
C:\msys64\mingw64\bin\python.exe %*
```

Which allows `system(['python.exe', args...])` and it's equivalent to `system(['C:\msys64\mingw64\bin\python.exe', args...])` or `MSYS2_PATH='C:\msys64\mingw64\bin\'; system(['%{MSYS2_PATH}python.exe', args...])`. If you want it to also inherit MSYS2's env:

```bat
@echo off
C:\msys64\mingw64\bin\bash.exe -leo pipefail -c python %*
```

capocasa commented:
As mentioned many times, I cannot influence what the system command does, becuse it is part of various software packages I have no control over. As it is, the system command does this:

system("autoreconf")

So I can put an autoreconf.bat in the path- that is *exactly* what I proposed in my guide. I have no idea what we are discussing.

eine commented:
As I said earlier:

> You need to clarify your scope. There are two different issues here. On the one hand, you want to run Unix tools installed through MSYS2's pacman in the cmd or powershell. On the other hand, you want to wrap MSYS2's shell in a custom entrypoint to bash.

> "Adding a corresponding .bat file that will run the interpreter" is exactly the same as "Providing a custom entrypoint to bash".

> Executing binaries directly is no issue because those are regular windows programs. The requirement of a *.bat/*.cmd applies only to executing bash/shell scripts without graphically opening an MSYS2 window.

> You forgot the most straightforward solution: call whichever Windows tool you want to use, from MSYS2's bash. If you open a bash shell first and then call any tool from there (any native Windows tool which might have been built and distributed completely outside of MSYS2), that one inherits the environment. Problem solved. Your tool will find all the Unix tools per se, without any tweak, any custom entrypoints...

If I ask you about an explicit example and you only reply "Python", then I cannot help with the other issue. How do you execute the software package that executes `system("autoreconf")` internally? Do execute that software package from MSYS2, or use a `*.bat`/`*.cmd` to do so. I provided an explicit example about how to execute `python` from `cmd.exe` and have a full featured MSYS2 Python with the whole MSYS2 env available.

Technically, it's the same solution. Conceptually it is not. You are suggesting to wrap each of the tools that a given software package uses internally. I'm suggesting to wrap a single tool, the execution of the given software package.

capocasa commented:
Ah, I see. I didn't mean to be rude, I'm just growing frustrated receiving the same advice over and over again even though I repeatedly told you it's infeasible. Please, if you want to help, work within the parameters I set out, they are there for a reason.

The command running it is a compiler called Nim. There would be too many issues and side effects to wrap it.

eine commented:
Nim is available on MSYS2 already: https://packages.msys2.org/package/mingw-w64-x86_64-nim. Maybe you can just use that instead of installing Nim otherwise?

capocasa commented:
> Nim is available on MSYS2 already: https://packages.msys2.org/package/mingw-w64-x86_64-nim. Maybe you can just use that instead of installing Nim otherwise?

I am making a package for nim. I cannot insist all users of my package install Nim through msys2, although I personally do that and like it. Again, if you want to help, please work within the parameters I set out, they are there for a reason.

eine commented:
The reference above is a Nim package. It feels strange to me that you can ask your users to install Unix tools through MSYS2 (because those are required dependencies) and require them to setup some custom manual tweaks; but it is unacceptable to install just an additional package and have all problems solved. At the same time, I can hardly follow the parameters that were set out, because they feel to be unnecessarily constrained up to being a niche/negligible use case. First, no specific tools were mentioned. Then, when I explained how to do it with Python, we switched autoreconf. Then, we switched to Nim and we added a new constraint. Please, put the practical (not abstract) use cases upfront. A reproducible script or CI workflow would be very helpful.

capocasa commented:
> unnecessarily constrained

That's a preconceived notion. I'm sorry, I can no longer reply now.

eine commented:
It's explained in the same message. Exactly the two previous sentences explain why that's my perception. It's therefore a falacy to portrait it as a preconception.

I'd prefer if you explained why users that want to build your package cannot use the Nim package of MSYS2, provided that they need to have MSYS2. I find it very different from any other Nim package which does not depend on MSYS2 to be built, so there is no condition for it to be installed.

However, there might be specific ABI conflicts which I'm not aware about. That would change my perception.

capocasa commented:
I'm sorry, I am unable to make arguments with you in a way that is economical with my resources. I told you what would be helpful to me, and you continue to ignore it. I can no longer reply now in this way. Please let it go, thank you for trying.

elieux commented:
Hey. I read the discussion here and I took what's been written into account for this response.

I believe advising people to permanently add their */usr/bin* to Windows PATH will cause more issues than it can possibly solve. Note that I'm not arguing against cross-invoking in general. If such behavior is required somewhere, I'd rather see a local solution like a *use-msys2.cmd* script that'd add */usr/bin* to PATH for that shell session only. People could use this script for their special use cases without causing side effects. Of course, running a proper MSYS2 session sounds better to me, but this response isn't about that.

Regarding shimbangs, I'd actually like to see something like this implemented, but more robust and covering the whole of MSYS2. See https://www.msys2.org/wiki/Devtopics/#fix-pacman-errors-wrt-conflicts-in-binfoo-vs-binfooexe for the original motivation and the idea.

While on the topic of "UNIX tools", there are various projects that try to provide them in a way that might be better suited for standalone use, like busybox-w32 or mksh/win32.

As for the provided documentation, thank you for the effort (including the rework and lengthy discussion), but in line with the above explanation, I won't accept it. I'd be okay publishing an appropriately annotated link pointing to your documentation, if you host it elsewhere, e.g. on Gist.
